### PR TITLE
Adapt integration checklist to GRUB as well

### PR DESCRIPTION
**Pull request by drewmoseley**
Signed-off-by: Drew Moseley <drew.moseley@northern.tech>

# External Contributor Checklist

🚨 Please review the [guidelines for contributing](https://github.com/mendersoftware/mender/blob/master/CONTRIBUTING.md) to this repository.

- [ ] Make sure that all commits are signed with [`git --signoff`](https://git-scm.com/book/en/v2/Git-Tools-Signing-Your-Work). Also note that the signoff author must match the author of the commit.

### Description

This is the basic set of changes needed so that the integration checklist also applies to GRUB. Basically, instead of relying on stopping at the U-Boot prompt and printing the environment, we do it solely from the shell prompt and verifying that bootcount is updated properly.  If updates in either direction are incorrect this will be detected by the bootcount variable.

Obviously this should get moved in the folder hierarchy but for now I just want to get the content reviewed.

### CODE REVIEW

File: 05.System-updates-Yocto-Project/02.Board-integration/02.Bootloader-support/02.U-Boot/02.Integration-checklist/docs.md [hunk 1]
```diff
@@ -60,111 +58,67 @@ This checklist will verify some key functionality aspects of the Mender integrat
...
 
-5. Verify that the two commands `fw_printenv` and `fw_setenv` are in the path and are executable. Calling them with no arguments will should give a variable listing and an error about missing variable name, respectively. This verifies that libubootenv tools are present on the device.
+5. Verify that the two commands `fw_printenv` and `fw_setenv` are in the path and are executable. Calling them with no arguments will should give a variable listing and an error about missing variable name, respectively. This verifies that the bootloader environment tools are present on the device.
 
-6. Now we will test that storing and reading values in the U-Boot environment works. Run:
+6. Now we will verify that Mender is running. Run the following:
 
    ```bash
-   fw_setenv mender_linux_test 1
+   pidof mender
```
Comment (lluiscampos) [inline 05.System-updates-Yocto-Project/02.Board-integration/02.Bootloader-support/02.U-Boot/02.Integration-checklist/docs.md:66]: This should be `pidof mender-client`

File: 05.System-updates-Yocto-Project/02.Board-integration/02.Bootloader-support/02.U-Boot/02.Integration-checklist/docs.md [hunk 2]
```diff
@@ -188,8 +142,18 @@ This checklist will verify some key functionality aspects of the Mender integrat
...
     ! The `mender_boot_part` number is the number of the first rootfs partition, and corresponds to the last component of the `MENDER_ROOTFS_PART_A` variable. If you've changed this variable in the Yocto configuration, you may need to use a different number.
 
-19. Reboot, but pull the power plug before the system has had time to finish booting. Mender will auto-commit the update if it is enabled as a service, which will defeat the purpose of this test, so it's important that the power is cut after the kernel has started booting, but before Mender has started.
+14. Reboot.  Now verify that the bootloader has updated the bootcount variable.
+
+    ```bash
+    fw_printenv bootcount
+    ```
+
+    This value should be returned as 1.  This step verifies that the bootloader and the bootloader environment tools can properly communicate and agree on the format of the bootloader environment.
```
Comment (lluiscampos) [inline 05.System-updates-Yocto-Project/02.Board-integration/02.Bootloader-support/02.U-Boot/02.Integration-checklist/docs.md:151]:
```suggestion
    This value should be returned as 1. This step verifies that the bootloader and the bootloader environment tools can properly communicate and agree on the format of the bootloader environment.
```

**Review by lluiscampos — CHANGES_REQUESTED** [review]
I found one typo, the rest looks good.

One general question, though. As this is now both U-Boot and GRUB, this integration checklist does not belong to `02.Board-integration/02.Bootloader-support/02.U-Boot/02.Integration-checklist` but maybe `02.Board-integration/02.Bootloader-support/02.Integration-checklist`?

**Review by kacf — COMMENTED** [review]
Nice changes, @drewmoseley!

Ironically, if we do [MEN-3978](https://tracker.mender.io/browse/MEN-3978) we may have to split this up again to avoid making the instructions too complicated. But we can cross that bridge when we get there.

Should this file be moved though? It is currently indexed under the U-Boot chapter.

**Comment by drewmoseley** [comment]
Yes, it should obviously be moved. I wanted to review the changes here first as it makes it more evident what specifically has changed.  I'll move it and update this PR and remove the RFC tag.

File: 05.System-updates-Yocto-Project/02.Board-integration/02.Bootloader-support/02.U-Boot/02.Integration-checklist/docs.md (hunk 1 above)
Comment (drewmoseley) [reply 05.System-updates-Yocto-Project/02.Board-integration/02.Bootloader-support/02.U-Boot/02.Integration-checklist/docs.md:66]: Are you sure? I know the service is named mender-client but I didn't think the executable was renamed.

**Comment by drewmoseley** [comment]
@mirzak is the "Client Installation" chapter a good place for this?

**Comment by mirzak** [comment]
As the instructions are Yocto specific, I would put it in,

```
05.System-updates-Yocto-Project/02.Board-integration/02.Bootloader-support/03-Integration-checklist/
```

**Comment by drewmoseley** [comment]
Hmm.  I wonder if we should make it more generic then.  But does it get too complicated that way?

File: 05.System-updates-Yocto-Project/02.Board-integration/02.Bootloader-support/02.U-Boot/02.Integration-checklist/docs.md (hunk 1 above)
Comment (lluiscampos) [reply 05.System-updates-Yocto-Project/02.Board-integration/02.Bootloader-support/02.U-Boot/02.Integration-checklist/docs.md]: Yes, you are completely right... I got confused.

**Comment by mirzak** [comment]
I guess the steps 1-3 are Yocto specific, the rest should be fairly generic.

Would this be equivalent:

```
# Remove kernel from rootfs A, this should result in a failure in case of the integration is trying to load the kernel from the wrong partition
rm /boot/uImage

fw_setenv mender_boot_part 3
fw_setenv mender_boot_part_hex 3
reboot
```

This is so that we can get rid of keeping track of timestamps of Linux kernel.

**Comment by drewmoseley** [comment]
Yeah I think that could work.  Let me write it up.

**Comment by drewmoseley** [comment]
@michaelatmender @lluiscampos @kacf please take another look.  I added a commit to remove the Yocto-specifics.

**Comment by drewmoseley** [comment]
@mirzak ping

File: 03.Client-installation/07.Integration-checklist/docs.md [hunk 3]
```diff
@@ -25,58 +25,25 @@ This checklist will verify some key functionality aspects of the Mender integrat
...
    ```bash
-   bitbake -c cleansstate virtual/kernel core-image-full-cmdline
+   mv /boot/uImage /boot/uImage.testing-backup
    ```
 
-   ! Note that `core-image-full-cmdline` should be switched to whatever image recipe you are testing with.
+   ! Note that `uImage` should be updated to match the kernel image type for your platform.
 
-   Then build the image normally.
+2. Verify that the two commands `fw_printenv` and `fw_setenv` are in the path and are executable. Calling them with no arguments will should give a variable listing and an error about missing variable name, respectively. This verifies that the bootloader environment tools are present on the device.
```
Comment (kacf) [inline 03.Client-installation/07.Integration-checklist/docs.md:34]: "...~will~ should..."

File: 03.Client-installation/07.Integration-checklist/docs.md [hunk 4]
```diff
@@ -108,17 +75,27 @@ This checklist will verify some key functionality aspects of the Mender integrat
...
+7. Repeat step 4, but this time verify that the root filesystem partition is different than the previous check. This verifies that both the root filesystem partition and kernel image are properly loaded from partition B.
 
-    !!! In the case of raw flash storage, the second rootfs was not updated, hence the build time listed in `/proc/version` will be the same as for the first partition.
+8. Now, restore the kernel image on partition A:
 
-12. Repeat step 8, but this time using `/dev/mmcblk0p3` (or `ubi0_1`). This verifies that the correct rootfs is mounted when partition B is active.
+    - When using SD card or eMMC storage:
+    ```bash
+    mount /dev/mmcblk2 /mnt
+    mv /boot/uImage.testing-backup /boot/uImage
```
Comment (kacf) [inline 03.Client-installation/07.Integration-checklist/docs.md:85]: I think you forgot `/mnt` in the paths here.

File: 03.Client-installation/07.Integration-checklist/docs.md [hunk 5]
```diff
@@ -108,17 +75,27 @@ This checklist will verify some key functionality aspects of the Mender integrat
...
+    umount /mnt
+    ```
 
-    ! As in the previous step, `/dev/mmcblk0p3` (or `ubi0_1`) may be different if `MENDER_ROOTFS_PART_B` or `MENDER_STORAGE_DEVICE` has been changed.
+    - When using raw flash storage:
+    ```bash
+    mount /dev/ubi0_1 /mnt
+    mv /boot/uImage.testing-backup /boot/uImage
```
Comment (kacf) [inline 03.Client-installation/07.Integration-checklist/docs.md:92]: Here as well.

File: 03.Client-installation/07.Integration-checklist/docs.md [hunk 6]
```diff
@@ -108,17 +75,27 @@ This checklist will verify some key functionality aspects of the Mender integrat
...
+    ```bash
+    mount /dev/mmcblk2 /mnt
+    mv /boot/uImage.testing-backup /boot/uImage
+    umount /mnt
+    ```
 
-    ! As in the previous step, `/dev/mmcblk0p3` (or `ubi0_1`) may be different if `MENDER_ROOTFS_PART_B` or `MENDER_STORAGE_DEVICE` has been changed.
+    - When using raw flash storage:
+    ```bash
+    mount /dev/ubi0_1 /mnt
```
Comment (kacf) [inline 03.Client-installation/07.Integration-checklist/docs.md:91]: Shouldn't this be `ubi0_0` if it's the first partition?

File: 03.Client-installation/07.Integration-checklist/docs.md [hunk 7]
```diff
@@ -25,58 +25,25 @@ This checklist will verify some key functionality aspects of the Mender integrat
 
 !!! Steps 1-4 can be skipped when verifying [raw flash](../../../03.Raw-flash/docs.md) integration.
 
-1. As part of the test, we will need two different Linux kernels, in order to verify that both are booted correctly when they should. Therefore, before building the images you will test with, run the commands:
+1. As part of the test, we will rename the Linux kernel in one of the partitions to ensure that the correct kernel image is loaded. In your initially booted system, do the following to hide the Linux kernel from the bootloader
 
    ```bash
-   bitbake -c cleansstate virtual/kernel core-image-full-cmdline
+   mv /boot/uImage /boot/uImage.testing-backup
```
Comment (kacf) [inline 03.Client-installation/07.Integration-checklist/docs.md:29]: Nice, a bit more straightforward than the original!

**Review by kacf — CHANGES_REQUESTED** [review]
Sorry, this one slipped.

**Comment by drewmoseley** [comment]
@kacf thanks for the review. I pushed a new commit with your recommendations.

**Review by kacf — APPROVED** [review]
Looks good! @lluiscampos: You good with this too?

I'm not sure about the passive voice checker which is failing now. Are these actually new words, or is it triggering on pre-existing content?

**Comment by mirzak** [comment]
The passive voice checker is enabled for the "Client installation" section, and since the integration checklist was moved there the script will run on the "new" content. 

Needs to be fixed up.

**Review by mirzak — APPROVED** [review]

**Review by lluiscampos — APPROVED** [review]
LGTM!

Sorry for lagging in the final review.

File: 03.Client-installation/07.Integration-checklist/docs.md [hunk 8]
```diff
@@ -123,11 +123,11 @@ This checklist will verify some key functionality aspects of the Mender integrat
     fw_printenv bootcount
     ```
 
-    This value should be returned as 1. This step verifies that the bootloader and the bootloader environment tools can properly communicate and agree on the format of the bootloader environment.
+    This should return the value 1. This step verifies that the bootloader and the bootloader environment tools can properly communicate and agree on the format of the bootloader environment.
 
 11. Repeat step 9.
 
-12. Reboot, but pull the power plug before the system has had time to finish booting. Mender will auto-commit the update if it is enabled as a service, which will defeat the purpose of this test, so it's important that the power is cut after the kernel has started booting, but before Mender has started.
+12. Reboot, but pull the power plug before the system has had time to finish booting. The Mender will service will auto-commit the update, which will defeat the purpose of this test. It ss important to reboot the system after the kernel has started booting, but before Mender has started.
```
Comment (kacf) [inline 03.Client-installation/07.Integration-checklist/docs.md:130]: "It ss"

**Review by kacf — APPROVED** [review]
One small typo below. You can just go ahead and merge afterwards. Looks good!

File: 03.Client-installation/07.Integration-checklist/docs.md (hunk 8 above)
Comment (drewmoseley) [reply 03.Client-installation/07.Integration-checklist/docs.md]: Fixed.

**Pull request merged**